### PR TITLE
session: new package for caching discv5 sessions

### PR DESCRIPTION
**Pull request by lechengfan**
Create a simple cache for storing recognized discv5 sessions in an LRU manner. 

Will be used whenever the node receives a request. Pseudocode as the following:

```
    // we receive message from a UDP endpoint
    session, found := cache.GetSession(Endpoint{NodeID: "something", Port: "something else"})
    if !found {
         //respond with WHOAREYOU challenge
    }
    // get handshake information and validate encrypted message
```

### CODE REVIEW

File: session/cache.go [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+}
+
+func (c *Cache) StoreSession(endpoint Endpoint, session interface{}) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	if el, ok := c.entriesMap[endpoint]; ok {
+		// cache hit; update
+		el.Value.(*cacheEntry).v = session
+		c.lru.MoveToFront(el)
```
Comment (ryandotsmith) [inline session/cache.go:48]: can you return here and then omit the `else {}`

File: session/cache.go [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+
+func NewCache() *Cache {
+	return &Cache{
+		lru:        list.New(),
+		size:       defaultMaxSize,
+		entriesMap: make(map[Endpoint]*list.Element),
+	}
+}
+
+func (c *Cache) StoreSession(endpoint Endpoint, session interface{}) {
```
Comment (ryandotsmith) [inline session/cache.go:41]: I think the `Session` suffix is a bit redundant. The call sites should be something like

```go
sessionCache.Store(endpoint, session)
```
and
```go
session, ok := sessionCache.Get(endpoint)
```

**Review by ryandotsmith — APPROVED** [review]
a few minor suggestions otherwise LGTM

File: session/cache.go (hunk 2 above)
Comment (lechengfan) [reply session/cache.go:41]: Will do!

**Pull request merged**